### PR TITLE
feat: Add SwanLab integration for experiment tracking

### PR DESCRIPTION
**Pull request by PraMamba**
## Overview
This PR adds SwanLab integration to Axolotl, enabling comprehensive experiment tracking and visualization for model training workflows.

## Features
- 🔌 Plugin-based integration following Axolotl's architecture
- 📊 Automatic metrics logging (loss, learning rate, etc.)
- ⚙️ Hyperparameter tracking
- 🌐 Cloud and local mode support
- 🎯 Custom callbacks for advanced use cases
- 🔔 Lark webhook integration for team notifications
- 📝 RLHF completion logging for DPO/KTO/ORPO/GRPO training
- ⏱️ Performance profiling utilities for trainer methods
- 📝 Comprehensive documentation and examples

## Changes
- New `SwanLabPlugin` in `src/axolotl/integrations/swanlab/`
- Configuration schema via `SwanLabArgs` with Lark and profiling support
- RLHF completion logging via `SwanLabRLHFCompletionCallback`
- Performance profiling utilities for measuring trainer execution times
- Test suite with comprehensive coverage (69 tests, 2,300+ lines)
- Integration with existing Axolotl callback system

## Files Added
- `src/axolotl/integrations/swanlab/__init__.py` - Plugin initialization
- `src/axolotl/integrations/swanlab/README.md` - Comprehensive documentation
- `src/axolotl/integrations/swanlab/args.py` - Configuration schema
- `src/axolotl/integrations/swanlab/plugins.py` - Plugin implementation
- `src/axolotl/integrations/swanlab/callbacks.py` - RLHF completion logging
- `src/axolotl/integrations/swanlab/completion_logger.py` - Completion management
- `src/axolotl/integrations/swanlab/profiling.py` - Performance profiling
- `src/axolotl/utils/callbacks/swanlab.py` - Basic callback
- `tests/integrations/test_swanlab.py` - Comprehensive test suite

## Testing
- ✅ Unit tests: `pytest tests/integrations/test_swanlab.py` (69 tests passed)
- ✅ Pre-commit checks: All checks passed (ruff, mypy, bandit)
- ✅ Type safety: Full mypy type annotations
- ✅ Code quality: Passed all linting rules

## Documentation
- Installation instructions in `src/axolotl/integrations/swanlab/README.md`
- Configuration examples and usage patterns
- API documentation with inline docstrings
- Integration guide for different training modes

## Usage Example
```yaml
plugins:
  - axolotl.integrations.swanlab.SwanLabPlugin

use_swanlab: true
swanlab_project: my-llm-finetuning
swanlab_experiment_name: llama3-lora-test
swanlab_mode: cloud  # or 'local', 'offline', 'disabled'

# Optional: Lark notifications
swanlab_lark_webhook_url: https://your-lark-webhook
swanlab_lark_secret: your-secret

# Optional: RLHF completion logging
swanlab_log_completions: true
swanlab_completion_log_interval: 100

# Optional: Performance profiling
swanlab_enable_profiling: true
swanlab_profiling_min_duration_ms: 1.0
```

## Related Issues
N/A - This is a new integration feature

## Checklist
- [x] Code follows Axolotl style guidelines
- [x] Tests pass locally
- [x] Pre-commit checks pass
- [x] Type annotations added (mypy compliant)
- [x] Documentation is complete
- [x] No breaking changes to existing functionality

## Summary by CodeRabbit

* **New Features**
  * Added SwanLab experiment tracking integration with support for cloud, local, and offline modes.
  * Added RLHF completion logging for DPO, KTO, ORPO, and GRPO trainers.
  * Added Lark (Feishu) notification support for training events.
  * Added performance profiling utilities for tracking execution metrics.
  * Added configurable SwanLab settings for project, experiment, API credentials, and logging behavior.

* **Documentation**
  * Added comprehensive integration guide covering installation, configuration, troubleshooting, and usage examples.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

**Comment by NanoCode012** [comment]
Thanks for the PR. We'll get back to you when the folks are back

File: src/axolotl/integrations/swanlab/args.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
+"""SwanLab configuration arguments"""
+
+from pydantic import BaseModel, Field, field_validator, model_validator
+
+
+class SwanLabConfig(BaseModel):
+    """SwanLab configuration subset"""
+
+    use_swanlab: bool | None = Field(
+        default=None,
```
Comment (winglian) [inline src/axolotl/integrations/swanlab/args.py]: The default can be set to `True` as including the SwanLab plugin is pretty explicit. This will help reduce boilerplate to enable the plugin.

File: examples/swanlab/dpo-swanlab-full-featured.yml [hunk 2]
```diff
@@ -0,0 +1,329 @@
...
+# - Lark (Feishu) team notifications
+# - Team workspace collaboration
+#
+# Use this as a reference for production RLHF training setups.
+#
+# To run:
+#   export SWANLAB_API_KEY=your-api-key
+#   export SWANLAB_LARK_WEBHOOK_URL=https://open.feishu.cn/...
+#   export SWANLAB_LARK_SECRET=your-webhook-secret
+#   accelerate launch -m axolotl.cli.train examples/swanlab/dpo-swanlab-full-featured.yml
```
Comment (winglian) [inline examples/swanlab/dpo-swanlab-full-featured.yml:16]: `axolotl train examples/swanlab/dpo-swanlab-full-featured.yml`

**Review by winglian — COMMENTED** [review]
Generally, everything seems 👍 . Just a few things to address from the coderabbit review and my comments and should be good to go.

**Comment by PraMamba** [comment]
@winglian 

All review feedback has been addressed! 🎉

## Changes Made

**Your feedback:**
- ✅ Changed `use_swanlab` default value to `True` (args.py:10)
- ✅ Verified example command is correct (dpo-swanlab-full-featured.yml:16)

**CodeRabbit Major:**
- ✅ Added buffer clearing after periodic logging to prevent duplicates (callbacks.py:163)

**CodeRabbit Minor:**
- ✅ Added safe exception handling in config fallback (plugins.py:430)
- ✅ Used context managers for file operations (swanlab.py:177, :223)

**CodeRabbit Nitpicks:**
- ✅ Sorted `__all__` alphabetically (__init__.py:6)
- ✅ Replaced 5× `LOG.error` with `LOG.exception` for better debugging
- ✅ Added language specifiers to 5 README code blocks

## Verification

All tests passing locally:
- ✅ **PyTest**: 69/69 tests passed
- ✅ **Pre-commit**: All hooks passed (ruff, mypy, bandit)
- ✅ **Git**: All changes committed (588f1123)

Ready for re-review! 🚀

Commit: https://github.com/axolotl-ai-cloud/axolotl/pull/3334/commits/588f11234b9285b8217e6044c732da9c642583c8

**Review by winglian — APPROVED** [review]
